crypto/signatures.go

```diff
@@ -69,8 +72,47 @@ func VerifyHash(data Hash, pk PublicKey, sig Signature) error {
 	return nil
 }
 
+// WriteSignedObject writes a length-prefixed object followed by its signature.
+func WriteSignedObject(w io.Writer, obj interface{}, sk SecretKey) error {
+	encObj := encoding.Marshal(obj)
+	sig, _ := SignHash(HashBytes(encObj), sk) // no error possible
+	return encoding.NewEncoder(w).EncodeAll(encObj, sig)
+}
+
+// ReadSignedObject reads a length-prefixed object followed by its signature,
+// and verifies the signature.
+func ReadSignedObject(r io.Reader, obj interface{}, maxLen uint64, pk PublicKey) error {
+	// read the encoded object and signature
+	var encObj []byte
+	var sig Signature
+	err := encoding.NewDecoder(r).DecodeAll(&encObj, &sig)
+	if err != nil {
+		return err
+	}
+	// verify the signature
+	if err := VerifyHash(HashBytes(encObj), pk, sig); err != nil {
+		return err
+	}
+	// decode the object
+	return encoding.Unmarshal(encObj, obj)
```

Comment:
I guess it doesn't really simplify decoding much, but still I worry that the other way would be confusing when writing protocols like negotiation. (I assumed sig was first, adjust accordingly if putting sig second)

```go
objAndSigBytes, err := ioutil.ReadAll(r)
objBytes := objAndSigBytes[SignatureSize:]
err = VerifyHash(objBytes), pk, objBytes)
return encoding.Unmarshal(objBytes, obj)
```